File: apps/backend/app/schemas/feature_selection.py
"""
Pydantic schemas for Feature Selection API endpoints.

Defines request/response models for feature selection operations,
supporting multiple selection algorithms with importance scoring.
"""

from pydantic import BaseModel, Field
from typing import List, Optional, Dict, Any, Literal
from datetime import datetime


# Enums for selection methods
SelectionMethod = Literal[
    "correlation",
    "mutual_info",
    "random_forest",
    "rfe",
    "lasso",
    "statistical"
]


class FeatureSelectionRequest(BaseModel):
    """Request schema for feature selection endpoint."""

    target_column: str = Field(
        ...,
        description="Name of the target column for supervised selection"
    )
    method: SelectionMethod = Field(
        ...,
        description="Feature selection algorithm to use"
    )
    top_k: Optional[int] = Field(
        None,
        ge=1,
        description="Number of top features to select (default: auto-determine)"
    )
    threshold: Optional[float] = Field(
        None,
        ge=0.0,
        le=1.0,
        description="Importance threshold for feature inclusion (0.0-1.0)"
    )
    correlation_threshold: Optional[float] = Field(
        0.7,
        ge=0.0,
        le=1.0,
        description="Threshold for detecting redundant features (default: 0.7)"
    )
    problem_type: Optional[Literal["classification", "regression"]] = Field(
        None,
        description="Type of ML problem (auto-detected if not provided)"
    )
    algorithm_params: Optional[Dict[str, Any]] = Field(
        default_factory=dict,
        description="Algorithm-specific parameters (e.g., n_estimators for Random Forest)"
    )
    sample_size: Optional[int] = Field(
        None,
        description="Sample size for large datasets (default: use full dataset)"
    )


class FeatureScore(BaseModel):
    """Feature with importance score."""

    feature_name: str = Field(..., description="Feature name")
    score: float = Field(..., description="Normalized importance score (0-1)")
    rank: int = Field(..., ge=1, description="Rank by importance (1=most important)")
    selected: bool = Field(..., description="Whether feature was selected")


class RedundantPair(BaseModel):
    """Pair of redundant features."""

    feature1: str = Field(..., description="First feature name")
    feature2: str = Field(..., description="Second feature name")
    correlation: float = Field(..., description="Correlation coefficient")
    recommendation: str = Field(
        ...,
        description="Recommendation for handling redundancy"
    )


class FeatureSelectionResult(BaseModel):
    """Response schema for feature selection endpoint."""

    dataset_id: str = Field(..., description="Dataset ID")
    method: SelectionMethod = Field(..., description="Selection method used")
    selected_features: List[str] = Field(
        ...,
        description="List of selected feature names"
    )
    feature_scores: List[FeatureScore] = Field(
        ...,
        description="All features with importance scores"
    )
    redundant_pairs: List[RedundantPair] = Field(
        default_factory=list,
        description="Pairs of highly correlated features"
    )
    correlation_matrix: Optional[Dict[str, Dict[str, float]]] = Field(
        None,
        description="Full correlation matrix for numeric features"
    )
    explanation: str = Field(
        ...,
        description="Human-readable explanation of selection results"
    )
    metadata: Dict[str, Any] = Field(
        default_factory=dict,
        description="Additional metadata (problem_type, n_samples, etc.)"
    )
    execution_time_ms: float = Field(
        ...,
        description="Execution time in milliseconds"
    )
    created_at: datetime = Field(
        default_factory=datetime.utcnow,
        description="Timestamp of selection"
    )


class FeatureImportanceRequest(BaseModel):
    """Request schema for feature importance calculation."""

    target_column: str = Field(..., description="Target column name")
    method: SelectionMethod = Field(..., description="Importance calculation method")
    problem_type: Optional[Literal["classification", "regression"]] = None
    algorithm_params: Optional[Dict[str, Any]] = Field(default_factory=dict)


class FeatureImportanceResponse(BaseModel):
    """Response schema for feature importance calculation."""

    dataset_id: str = Field(..., description="Dataset ID")
    method: SelectionMethod = Field(..., description="Method used")
    feature_scores: List[FeatureScore] = Field(
        ...,
        description="Features ranked by importance"
    )
    execution_time_ms: float = Field(..., description="Execution time")


class RedundancyDetectionRequest(BaseModel):
    """Request schema for redundancy detection."""

    correlation_threshold: float = Field(
        0.7,
        ge=0.0,
        le=1.0,
        description="Correlation threshold for redundancy"
    )
    method: Literal["pearson", "spearman", "kendall"] = Field(
        "pearson",
        description="Correlation method to use"
    )


class RedundancyDetectionResponse(BaseModel):
    """Response schema for redundancy detection."""

    dataset_id: str = Field(..., description="Dataset ID")
    redundant_pairs: List[RedundantPair] = Field(
        ...,
        description="Redundant feature pairs"
    )
    correlation_matrix: Dict[str, Dict[str, float]] = Field(
        ...,
        description="Full correlation matrix"
    )
    recommendations: List[str] = Field(
        default_factory=list,
        description="Recommendations for handling redundancy"
    )


class MethodComparisonRequest(BaseModel):
    """Request schema for comparing selection methods."""

    target_column: str = Field(..., description="Target column name")
    methods: List[SelectionMethod] = Field(
        ...,
        min_length=2,
        max_length=6,
        description="Methods to compare (2-6)"
    )
    top_k: Optional[int] = Field(
        10,
        ge=1,
        description="Number of top features per method"
    )
    problem_type: Optional[Literal["classification", "regression"]] = None


class MethodComparisonResult(BaseModel):
    """Results for a single method in comparison."""

    method: SelectionMethod = Field(..., description="Selection method")
    selected_features: List[str] = Field(..., description="Selected features")
    top_features: List[FeatureScore] = Field(
        ...,
        description="Top features with scores"
    )
    execution_time_ms: float = Field(..., description="Execution time")


class MethodComparisonResponse(BaseModel):
    """Response schema for method comparison."""

    dataset_id: str = Field(..., description="Dataset ID")
    results: List[MethodComparisonResult] = Field(
        ...,
        description="Results for each method"
    )
    consensus_features: List[str] = Field(
        ...,
        description="Features selected by all methods"
    )
    overlap_matrix: Dict[str, Dict[str, int]] = Field(
        ...,
        description="Overlap count between methods"
    )
    recommendations: str = Field(
        ...,
        description="Recommendations based on comparison"
    )


class SelectedFeaturesResponse(BaseModel):
    """Response schema for retrieving previously selected features."""

    dataset_id: str = Field(..., description="Dataset ID")
    selected_features: List[str] = Field(
        default_factory=list,
        description="List of previously selected feature names"
    )
    has_selection: bool = Field(
        ...,
        description="Whether any features have been selected"
    )

File: apps/frontend/lib/services/featureSelection.ts
/**
 * Feature Selection service for ML feature engineering
 *
 * Provides methods for feature selection operations including
 * multiple selection algorithms, importance calculation, and
 * redundancy detection.
 *
 * @module lib/services/featureSelection
 */

const API_BASE_URL = process.env.NEXT_PUBLIC_API_URL || 'http://localhost:8000/api/v1'

// =============================================================================
// Types
// =============================================================================

export type SelectionMethod =
  | 'correlation'
  | 'mutual_info'
  | 'random_forest'
  | 'rfe'
  | 'lasso'
  | 'statistical'

export type ProblemType = 'classification' | 'regression'

export interface FeatureScore {
  feature_name: string
  score: number
  rank: number
  selected: boolean
}

export interface RedundantPair {
  feature1: string
  feature2: string
  correlation: number
  recommendation: string
}

export interface FeatureSelectionRequest {
  target_column: string
  method: SelectionMethod
  top_k?: number
  threshold?: number
  correlation_threshold?: number
  problem_type?: ProblemType
  algorithm_params?: Record<string, any>
  sample_size?: number
}

export interface FeatureSelectionResult {
  dataset_id: string
  method: SelectionMethod
  selected_features: string[]
  feature_scores: FeatureScore[]
  redundant_pairs: RedundantPair[]
  correlation_matrix?: Record<string, Record<string, number>>
  explanation: string
  metadata: {
    problem_type: string
    n_samples: number
    n_features_original: number
    n_features_selected: number
    target_column: string
  }
  execution_time_ms: number
  created_at: string
}

export interface FeatureImportanceRequest {
  target_column: string
  method: SelectionMethod
  problem_type?: ProblemType
  algorithm_params?: Record<string, any>
}

export interface FeatureImportanceResponse {
  dataset_id: string
  method: SelectionMethod
  feature_scores: FeatureScore[]
  execution_time_ms: number
}

export interface RedundancyDetectionRequest {
  correlation_threshold?: number
  method?: 'pearson' | 'spearman' | 'kendall'
}

export interface RedundancyDetectionResponse {
  dataset_id: string
  redundant_pairs: RedundantPair[]
  correlation_matrix: Record<string, Record<string, number>>
  recommendations: string[]
}

export interface MethodComparisonRequest {
  target_column: string
  methods: SelectionMethod[]
  top_k?: number
  problem_type?: ProblemType
}

export interface MethodComparisonResult {
  method: SelectionMethod
  selected_features: string[]
  top_features: FeatureScore[]
  execution_time_ms: number
}

export interface MethodComparisonResponse {
  dataset_id: string
  results: MethodComparisonResult[]
  consensus_features: string[]
  overlap_matrix: Record<string, Record<string, number>>
  recommendations: string
}

export interface SelectedFeaturesResponse {
  dataset_id: string
  selected_features: string[]
  has_selection: boolean
}

// =============================================================================
// Service Class
// =============================================================================

export class FeatureSelectionService {
  private static async getHeaders(token: string | null): Promise<HeadersInit> {
    return {
      'Content-Type': 'application/json',
      ...(token && { Authorization: `Bearer ${token}` })
    }
  }

  /**
   * Run feature selection on a dataset
   *
   * Calls: POST /api/datasets/{datasetId}/features/select
   *
   * @param datasetId - Dataset identifier
   * @param request - Feature selection configuration
   * @param token - Authentication token
   * @returns Feature selection results with scores and explanations
   */
  static async selectFeatures(
    datasetId: string,
    request: FeatureSelectionRequest,
    token: string | null
  ): Promise<FeatureSelectionResult> {
    const response = await fetch(
      `${API_BASE_URL}/datasets/${datasetId}/features/select`,
      {
        method: 'POST',
        headers: await this.getHeaders(token),
        body: JSON.stringify(request)
      }
    )

    if (!response.ok) {
      const error = await response.json()
      throw new Error(error.detail || 'Feature selection failed')
    }

    return response.json()
  }

  /**
   * Calculate feature importance scores
   *
   * Calls: POST /api/datasets/{datasetId}/features/importance
   *
   * @param datasetId - Dataset identifier
   * @param request - Importance calculation configuration
   * @param token - Authentication token
   * @returns Feature importance scores ranked by importance
   */
  static async calculateImportance(
    datasetId: string,
    request: FeatureImportanceRequest,
    token: string | null
  ): Promise<FeatureImportanceResponse> {
    const response = await fetch(
      `${API_BASE_URL}/datasets/${datasetId}/features/importance`,
      {
        method: 'POST',
        headers: await this.getHeaders(token),
        body: JSON.stringify(request)
      }
    )

    if (!response.ok) {
      const error = await response.json()
      throw new Error(error.detail || 'Importance calculation failed')
    }

    return response.json()
  }

  /**
   * Detect redundant (highly correlated) features
   *
   * Calls: POST /api/datasets/{datasetId}/features/redundancy
   *
   * @param datasetId - Dataset identifier
   * @param request - Redundancy detection configuration
   * @param token - Authentication token
   * @returns Redundant feature pairs with correlations and recommendations
   */
  static async detectRedundancy(
    datasetId: string,
    request: RedundancyDetectionRequest,
    token: string | null
  ): Promise<RedundancyDetectionResponse> {
    const response = await fetch(
      `${API_BASE_URL}/datasets/${datasetId}/features/redundancy`,
      {
        method: 'POST',
        headers: await this.getHeaders(token),
        body: JSON.stringify(request)
      }
    )

    if (!response.ok) {
      const error = await response.json()
      throw new Error(error.detail || 'Redundancy detection failed')
    }

    return response.json()
  }

  /**
   * Compare multiple feature selection methods
   *
   * Calls: POST /api/datasets/{datasetId}/features/compare
   *
   * @param datasetId - Dataset identifier
   * @param request - Method comparison configuration
   * @param token - Authentication token
   * @returns Comparison results with consensus features and overlap matrix
   */
  static async compareMethods(
    datasetId: string,
    request: MethodComparisonRequest,
    token: string | null
  ): Promise<MethodComparisonResponse> {
    const response = await fetch(
      `${API_BASE_URL}/datasets/${datasetId}/features/compare`,
      {
        method: 'POST',
        headers: await this.getHeaders(token),
        body: JSON.stringify(request)
      }
    )

    if (!response.ok) {
      const error = await response.json()
      throw new Error(error.detail || 'Method comparison failed')
    }

    return response.json()
  }

  /**
   * Get previously selected features for a dataset
   *
   * Calls: GET /api/datasets/{datasetId}/features/selected
   *
   * @param datasetId - Dataset identifier
   * @param token - Authentication token
   * @returns Previously selected features or empty list
   */
  static async getSelectedFeatures(
    datasetId: string,
    token: string | null
  ): Promise<SelectedFeaturesResponse> {
    const response = await fetch(
      `${API_BASE_URL}/datasets/${datasetId}/features/selected`,
      {
        method: 'GET',
        headers: await this.getHeaders(token)
      }
    )

    if (!response.ok) {
      const error = await response.json()
      throw new Error(error.detail || 'Failed to retrieve selected features')
    }

    return response.json()
  }
}

// =============================================================================
// Helper Functions
// =============================================================================

/**
 * Get human-readable name for selection method
 */
export function getMethodName(method: SelectionMethod): string {
  const names: Record<SelectionMethod, string> = {
    correlation: 'Correlation',
    mutual_info: 'Mutual Information',
    random_forest: 'Random Forest',
    rfe: 'Recursive Feature Elimination',
    lasso: 'LASSO',
    statistical: 'Statistical Tests'
  }
  return names[method]
}

/**
 * Get description for selection method
 */
export function getMethodDescription(method: SelectionMethod): string {
  const descriptions: Record<SelectionMethod, string> = {
    correlation:
      'Measures linear relationships between features and target using Pearson correlation',
    mutual_info:
      'Captures non-linear dependencies using mutual information theory',
    random_forest:
      'Uses tree-based importance from Random Forest models, handles feature interactions',
    rfe:
      'Recursively removes features to find the optimal subset using backward elimination',
    lasso:
      'Uses L1 regularization to automatically select features while training linear models',
    statistical:
      'Fast statistical tests (Chi-squared for classification, F-test for regression)'
  }
  return descriptions[method]
}

/**
 * Get recommended use case for method
 */
export function getMethodUseCase(method: SelectionMethod): string {
  const useCases: Record<SelectionMethod, string> = {
    correlation: 'Best for: Quick analysis, linear relationships, interpretable results',
    mutual_info: 'Best for: Non-linear data, categorical variables, complex relationships',
    random_forest: 'Best for: Tree-based models, feature interactions, robust selection',
    rfe: 'Best for: Finding optimal feature subset, high accuracy requirements',
    lasso: 'Best for: Linear models, automatic regularization, sparse solutions',
    statistical: 'Best for: Large datasets, quick filtering, statistical validation'
  }
  return useCases[method]
}

/**
 * Format execution time for display
 */
export function formatExecutionTime(ms: number): string {
  if (ms < 1000) {
    return `${Math.round(ms)}ms`
  } else if (ms < 60000) {
    return `${(ms / 1000).toFixed(1)}s`
  } else {
    const minutes = Math.floor(ms / 60000)
    const seconds = Math.round((ms % 60000) / 1000)
    return `${minutes}m ${seconds}s`
  }
}
